### DISCUSSION TITLE
Move [...] import [...] into a type checking block.

### DISCUSSION THREAD

**Discussion by RomainBrault** [beartype/beartype#460] [Q&A] [answered]
When importing 

```python
 from __future__ import annotations
```

to be able to use | instead of Union[...] I get [TCH003](https://docs.astral.sh/ruff/rules/typing-only-standard-library-import/#typing-only-standard-library-import-tch003) warning raise by ruff/flake/etc. 

In other words: "Move [...] import [...] into a type checking block."

But using TYPE_CHECKING seems to break runtime type checking. So should I just ignore this rule and consider is spurious when doing dynamic type checking?

I think many people use TYPE_CHECKING to reduce import time while using static type checker. Moreover this warnings by ruff and flake maybe explain why so many library do so and why beartype have trouble ? (e.g. xarray, pydantic, etc.)

Note:
It is possible to configure `runtime-evaluated-base-classes = [...]`. But this is annoying to configure, especially when `beartype.claw` is so convenient to type check all :). Moreover many people will probably just follow the linter recommendation and ignore that it will be impossible to dynamically type-check their code...

**Comment by RomainBrault** [comment]
Okay I discovered some answers in https://github.com/pydata/xarray/issues/9581

**Comment by leycec** [comment, marked as answer]
Further confirmation that **Ruff is fundamentally insane,** huh? [Literally *everything* the Ruff documentation claims at `TCH003` is unsubstantiated bald-faced lies that harm the CPython runtime](https://docs.astral.sh/ruff/rules/typing-only-standard-library-import/#typing-only-standard-library-import-tch003). Notably, this:

> Unused imports add a performance overhead at runtime...

**Negligible.** CPython trivially caches imported modules at `sys.modules`. Ruff, you are spouting nonsensical gibberish to support your spurious claims – *and we ain't havin' none of it.*

> If an import is only used in typing-only contexts, it can instead be imported conditionally under an if `TYPE_CHECKING:` block to minimize runtime overhead.

**No.** There is *no* such thing as "typing-only contexts." Due to type aliases (e.g., `MuhTypeAlias = list[str]`), literally *any* attribute can be leveraged by *any* external package as a type hint somewhere. Differentiating between attributes referenced only as type hints from attributes referenced both as type hints and as non-type hints is infeasible in the general case. Syntactic rules in the CPython grammar that Ruff developers seem to think are "typing-only contexts" are *not* "typing-only," because runtime code (including CPython itself) introspects type hints produced by "typing-only contexts" to solve computational decision problems at runtime.

There are no "typing-only contexts." There is only the CPython runtime. Increasingly, CPython itself is beginning to perform runtime type-checking. Inappropriate use of `TYPE_CHECKING` threatens to semantically break CPython itself.

Consider [PEP 649 – Deferred Evaluation Of Annotations Using Descriptors](https://peps.python.org/pep-0649/), recently adopted by CPython 3.14. Under PEP 649, CPython ≥ 3.14 dynamically resolves forward references in type hints at runtime. Something that only runtime type-checkers like @beartype and `typeguard` once did, CPython itself will now be doing. CPython cannot dynamically resolve forward references to types isolated to `TYPE_CHECKING` blocks; third-party code doing so turns the `__annotations__` dunder attribute into an "exception bomb" that raises exceptions when introspecting type hints at runtime. Ergo, `TYPE_CHECKING` encourages PEP 649 violations that break the CPython runtime itself.

This is the most destructive advice that anyone has ever given in the history of humanity.

> If a *class* requires that type annotations be available at runtime...

**CPython itself requires type hints at runtime.** CPython itself is *not* a "class." Neither are hybrid runtime-static type-checkers, which can be trivially applied to an entire application stack at runtime (e.g., via the `beartype_all()` import hook). All third-party packages thus need to expose their type hints at runtime. It's not up to individual third-party packages whether their type hints should be available at runtime or not. Type hints should *always* be available at runtime. Any other approach violates core PEP standards, breaks CPython itself, and breaks all runtime type-checkers.

Go away, Ruff! You have no idea of what you pontificate.

![](https://media1.tenor.com/m/PcUqTFgnNdEAAAAd/angry-kitty.gif)
<sup>*@beartype is squinting at you, Ruff*</sup>

**Comment by RomainBrault** [comment]
Thanks for the answer!

I would like to nuance a bit the usage of types at runtime. This can be seen as a waste in _some_ programs that use type checking solely for annotation purpose (and static type checking). For this programs types are indeed "useless" at runtime in the sense that they do not modify the behavior of the app/library of the author. So they are just seen as wasted import time (which sadly can be quite high with Python).

So I think beyond ruff's/flake recommendations, people will want to use TYPE_CHECKING for this reason as typing seems to be a waste of computation time (initial import time mainly) for only static usage.

Note:

You may have answered here:

> CPython itself requires type hints at runtime. CPython itself is not a "class." Neither are hybrid runtime-static type-checkers, which can be trivially applied to an entire application stack at runtime (e.g., via the beartype_all() import hook). All third-party packages thus need to expose their type hints at runtime. It's not up to individual third-party packages whether their type hints should be available at runtime or not. Type hints should always be available at runtime. Any other approach violates core PEP standards, breaks CPython itself, and breaks all runtime type-checkers.

However I am not sure I understand how it could break CPython as we are still supposed to be able to use untyped Python?

Meanwhile I'm gonna spend some time on [PEP 649](https://peps.python.org/pep-0649/) :smiley:

**Reply by leycec** [reply]
> For this programs types are indeed "useless" at runtime in the sense that they do not modify the behavior of the app/library of the author.

*...heh.* I get what you are saying. That's the Old World™ view of type hints, right? That's still the most popular conception of type hints. In the Old World™, type hints were only used at runtime when an author explicitly intended those type hints to be used at runtime (e.g., by decorating a callable with `@beartype`). It was thus reasonable to "optimize away" type hints *not* intended to be used at runtime via `TYPE_CHECKING` blocks.

Two modern trends upend that Old World™ mentality. Welcome to the New World™, where the Old World™ sucks and is considered harmful. :laughing: 

## Hybrid Runtime-static Type-checking Import Hooks

`beartype_all()`, `beartype_package()`, `beartype_packages()` operate on *any* arbitrary packages. If your package is one of those packages *and* if your package has prematurely optimized away type hints via `TYPE_CHECKING` blocks, your package is now **runtime-hostile** (i.e., incompatible with those import hooks).

You may not care about this as a package developer, but your users probably do. Do @beartype users care that increasingly many upstream packages are runtime-hostile? They do. They are complaining on the @beartype issue tracker about runtime-hostile packages on a daily basis. I've had to field like *twenty gazillion* issues over the past two weeks about this exact same topic. I've had to respond by **internally blacklisting** two popular third-party packages that are runtime-hostile: `xarray` and Pydantic. Yes, *the* Pydantic. Let's not ask why a runtime type-checker like Pydantic is itself runtime-hostile.

This is now @beartype's issue. Inappropriate assumptions made by other package developers have made this @beartype's issue without our consent. I sigh! I sigh so hard and then I faceplant onto the bed. :face_exhaling: 

## Python 3.14: PEP 649 Means That Runtime-Hostile Code Will Explode

This is PEP 649:

```python
def muh_func(muh_forward_ref: MuhUndefinedClass): pass
class MuhUndefinedClass(object): pass
```

See what magic we did there? The `muh_func` function is annotated by a type that has yet to be defined! The above code raises exceptions under CPython < 3.14 but works under CPython ≥ 3.14. PEP 649 is the dark magic that makes this worky.

PEP 649 is... *intense.* Basically, PEP 649 means that CPython ≥ 3.14 will now be doing something like what @beartype is currently doing. Specifically, CPython ≥ 3.14 will be dynamically replacing type hints that refer to undefined types with **forward reference proxies** (i.e., low-level C-based objects that transparently proxy `isinstance()` and `issubclass()` calls against undefined types by deferring the dynamic importation of those types to the last possible moment).

Here's where Ruff enters the picture and destroys everything. If a package follows Ruff's insane runtime-hostile advice to prematurely optimize away type hints under CPython ≥ 3.14, what happens? Consider this code snippet lifted directly from the Ruff TCH003 documentation:

```python
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from pathlib import Path


def func(path: Path) -> str:
    return str(path)
```

Recall that `TYPE_CHECKING` evaluates to `False` at runtime. Also note that `from __future__ import annotations` reduces to a noop under CPython ≥ 3.14. That statement no longer does anything or means anything. So, the above reduces to:

```python
def func(path: Path) -> str:
    return str(path)
```

Crucially, **Ruff's advice has been rendered insane.** Do you see? `Path` is undefined at runtime. Moreover, the package that `Path` resides in (i.e., `pathlib`) has been optimized away from runtime introspection. So what happens when somebody tries to introspect the type hints of the above `func` function by accessing the `func.__annotations__` dunder dictionary? Previously, that "just worked." You at least got a string back. It may not have been a *usable* string – but at least you had a string. You could do something with even an usable string (e.g., log a warning, raise an exception).

Now, however, the following code snippet that previously "worked" explodes horribly under Python ≥ 3.14:

```python
def func(path: Path) -> str:
    return str(path)

print(func.__annotations__['path'])  # <-- this blows up horribly

# This also blows up horribly, even though we're trying *MUCH*
# harder to safely get this type hint with the PEP 649 API.
import inspect
print(inspect.get_annotations(func, format=inspect.FORWARDREF)['path'])  # <-- even this blows up
```

This isn't CPython 3.14's fault. This isn't PEP 649's fault. This is entirely the fault of third-party static linters like Ruff that have no idea what they're doing. They don't. Ruff developers are *not* CPython developers. You should believe what CPython developers say and only what CPython developers say. Conversely, you should *not* believe what Ruff developers say.

Of course, I'm not a CPython developer either. By the above logic, you should ignore everything I say.

![](https://media1.tenor.com/m/xddGvs4_orcAAAAC/waynes-world-no.gif)
<sup>*gods, what am I saying!?!?*</sup>

**Comment by leycec** [comment]
**...lol.** I neglected to actually explain "How to type-check circular imports with @beartype." I kinda just ranted and talked at you a lot instead, @RomainBrault. It must be that time of year. :sweat_smile:

Happily, type-checking circular imports is a **trivial one-liner** with @beartype. You don't have to import *anything*; indeed, you usually shouldn't import anything. That's the whole point of circular imports, right? Here we go. Blink and you'll miss it:

```python
def muh_func(muh_arg: 'some_package.some_submodule.SomeClass'): ...
```

And... *that's it.* Seriously. Avoid circular imports by annotating circular types with absolute forward references. Supah-easy. Note the lack of `import` statements. This "just works" out-of-the-box with @beartype. I should probably write a new FAQ entry about this at some point. Until then, all I have are incoherent GitHub comments. `</sigh>`

Have a great weekend! Let's get ready for Canadian Thanksgiving, everybody. :turkey: :axe: :plate_with_cutlery:
